zooniverse/subject-set-search-api
# Community Catalog: add Projects database to Datasette

## PR Overview

Related repo: https://github.com/zooniverse/community-catalog
Related PR: #45

This PR adds the `projects` database to the Datasette service.

Intended functionality:
- The new `projects` database should exist side-by-side with the existing `subjects` database. 
  - For this PR, the projects DB will have only one project with 10 subjects.
- Each table in the `projects` database contains _every Subject_ from a pre-selected set of projects.
  - The "projects config" (see `PROJECTS` object in projects.js) lists these pre-selected projects _and_ the metadata fields we want to pull from that project's subjects.

Code changes:
- New script: `projects.js` pulls every Subject from a list of Projects, then creates a new Datasette-ready database for that project.
- Build changes:
  - Dockerfile now pre-builds `data/projects` sub-folder. (See #45 for the new subfolder additions)
  - import-csv-files-to-sqlite.sh now also processes CSVs created by projects.js into projects.db
  - Datasette now serves `projects.db` in immutable mode (see Dockerfile and docker-compose.yaml)

### Testing

Setup:
- Build and run the project in Docker.
- Open http://127.0.0.1:8001/

Testing Steps:
- Baseline: ensure that `subjects` database continues to exist.
- New: check that new `projects` database exists.
  - `projects` db should have exactly 1 table (proj_21084) with 10 rows

### Status

Ready for review!

Next PR: add Scarlets & Blues to the PROJECTS config, to test performance.
